From: Karthikeyan Periyasamy <quic_periyasa@quicinc.com>
Date: Wed, 22 Mar 2023 12:29:43 +0530
Subject: [PATCH] wifi: ath12k: Fetch DA mcbc from msdu end instead of mpdu

Avoid the MPDU start rx tlv dependency for fetching DA Multicast/Broadcast param
since MPDU is valid only on the MPDU boundary packets also it gives the RA is mcbc.
So fetch this information from MSDU_END rx tlv since its valid for each complete
msdu packet. Also rename the existing mpdu mcbc as RA prespective and implement
msdu_end mcbc as DA prespective.

Signed-off-by: Karthikeyan Periyasamy <quic_periyasa@quicinc.com>
---
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -205,11 +205,11 @@ static void ath12k_dp_rxdesc_set_msdu_le
 	ab->hw_params->hal_ops->rx_desc_set_msdu_len(desc, len);
 }
 
-static bool ath12k_dp_rx_h_is_mcbc(struct ath12k_base *ab,
-				   struct hal_rx_desc *desc)
+static bool ath12k_dp_rx_ra_is_mcbc(struct ath12k_base *ab,
+				    struct hal_rx_desc *desc)
 {
 	return (ath12k_dp_rx_h_first_msdu(ab, desc) &&
-	        ab->hw_params->hal_ops->rx_desc_is_mcbc(desc));
+		ab->hw_params->hal_ops->rx_desc_ra_is_mcbc(desc));
 }
 
 static bool ath12k_dp_rxdesc_mac_addr2_valid(struct ath12k_base *ab,
@@ -255,6 +255,12 @@ static inline u8 ath12k_dp_rx_get_msdu_s
 	return ab->hw_params->hal_ops->rx_desc_get_msdu_src_link_id(desc);
 }
 
+static bool ath12k_dp_rx_da_is_mcbc(struct ath12k_base *ab,
+				    struct hal_rx_desc *desc)
+{
+	return ab->hw_params->hal_ops->rx_desc_da_is_mcbc(desc);
+}
+
 static void ath12k_dp_service_mon_ring(struct timer_list *t)
 {
 	struct ath12k_base *ab = from_timer(ab, t, mon_reap_timer);
@@ -2754,8 +2760,7 @@ static void ath12k_dp_rx_h_mpdu(struct a
 	tid = ath12k_dp_rx_h_tid(ab, rx_desc);
 	/* PN for multicast packets will be checked in mac80211 */
 	rxcb = ATH12K_SKB_RXCB(msdu);
-	fill_crypto_hdr = ath12k_dp_rx_h_is_mcbc(ar->ab, rx_desc);
-	rxcb->is_mcbc = fill_crypto_hdr;
+	rxcb->is_mcbc = ath12k_dp_rx_da_is_mcbc(ar->ab, rx_desc);
 
 	if (rxcb->is_mcbc)
 		rxcb->peer_id = ath12k_dp_rx_h_peer_id(ar->ab, rx_desc);
@@ -2832,6 +2837,7 @@ static void ath12k_dp_rx_h_mpdu(struct a
 	if (err_bitmap & HAL_RX_MPDU_ERR_TKIP_MIC)
 		rx_status->flag |= RX_FLAG_MMIC_ERROR;
 
+	fill_crypto_hdr = ath12k_dp_rx_ra_is_mcbc(ar->ab, rx_desc);
 	if (is_decrypted) {
 		rx_status->flag |= RX_FLAG_DECRYPTED | RX_FLAG_MMIC_STRIPPED;
 
--- a/drivers/net/wireless/ath/ath12k/hal.c
+++ b/drivers/net/wireless/ath/ath12k/hal.c
@@ -412,10 +412,10 @@ static u8 *ath12k_hw_qcn9274_rx_desc_mpd
 	return desc->u.qcn9274.mpdu_start.addr2;
 }
 
-static bool ath12k_hw_qcn9274_rx_desc_is_mcbc(struct hal_rx_desc *desc)
+static bool ath12k_hw_qcn9274_rx_desc_ra_is_mcbc(struct hal_rx_desc *desc)
 {
 	return __le32_to_cpu(desc->u.qcn9274.mpdu_start.info6) &
-	       RX_MPDU_START_INFO6_MCAST_BCAST;
+			     RX_MPDU_START_INFO6_MCAST_BCAST;
 }
 
 static void ath12k_hw_qcn9274_rx_desc_get_dot11_hdr(struct hal_rx_desc *desc,
@@ -505,6 +505,12 @@ static u8 ath12k_hw_qcn9274_rx_desc_get_
 			     RX_MSDU_END_64_TLV_SRC_LINK_ID);
 }
 
+static bool ath12k_hw_qcn9274_rx_desc_da_is_mcbc(struct hal_rx_desc *desc)
+{
+	return le16_get_bits(desc->u.qcn9274.msdu_end.info5,
+			     RX_MSDU_END_INFO5_DA_IS_MCBC);
+}
+
 static int ath12k_hal_srng_create_config_qcn9274(struct ath12k_base *ab)
 {
 	struct ath12k_hal *hal = &ab->hal;
@@ -698,7 +704,7 @@ const struct hal_ops hal_qcn9274_ops = {
 	.rx_desc_get_msdu_end_offset = ath12k_hw_qcn9274_rx_desc_get_msdu_end_offset,
 	.rx_desc_mac_addr2_valid = ath12k_hw_qcn9274_rx_desc_mac_addr2_valid,
 	.rx_desc_mpdu_start_addr2 = ath12k_hw_qcn9274_rx_desc_mpdu_start_addr2,
-	.rx_desc_is_mcbc = ath12k_hw_qcn9274_rx_desc_is_mcbc,
+	.rx_desc_ra_is_mcbc = ath12k_hw_qcn9274_rx_desc_ra_is_mcbc,
 	.rx_desc_get_dot11_hdr = ath12k_hw_qcn9274_rx_desc_get_dot11_hdr,
 	.rx_desc_get_crypto_header = ath12k_hw_qcn9274_rx_desc_get_crypto_hdr,
 	.create_srng_config = ath12k_hal_srng_create_config_qcn9274,
@@ -712,6 +718,7 @@ const struct hal_ops hal_qcn9274_ops = {
 	.rxdma_ring_wmask_rx_mpdu_start = ath12k_hal_rx_mpdu_start_wmask_get,
 	.rxdma_ring_wmask_rx_msdu_end = ath12k_hal_rx_msdu_end_wmask_get,
 	.rx_desc_get_msdu_src_link_id = ath12k_hw_qcn9274_rx_desc_get_msdu_src_link,
+	.rx_desc_da_is_mcbc = ath12k_hw_qcn9274_rx_desc_da_is_mcbc,
 };
 
 static bool ath12k_hw_wcn7850_rx_desc_get_first_msdu(struct hal_rx_desc *desc)
@@ -880,10 +887,10 @@ static u8 *ath12k_hw_wcn7850_rx_desc_mpd
 	return desc->u.wcn7850.mpdu_start.addr2;
 }
 
-static bool ath12k_hw_wcn7850_rx_desc_is_mcbc(struct hal_rx_desc *desc)
+static bool ath12k_hw_wcn7850_rx_desc_ra_is_mcbc(struct hal_rx_desc *desc)
 {
 	return __le32_to_cpu(desc->u.wcn7850.mpdu_start.info6) &
-	       RX_MPDU_START_INFO6_MCAST_BCAST;
+			     RX_MPDU_START_INFO6_MCAST_BCAST;
 }
 
 static void ath12k_hw_wcn7850_rx_desc_get_dot11_hdr(struct hal_rx_desc *desc,
@@ -1122,6 +1129,12 @@ static u32 ath12k_hw_wcn7850_dp_rx_h_mpd
 	return errmap;
 }
 
+static bool ath12k_hw_wcn7850_rx_desc_da_is_mcbc(struct hal_rx_desc *desc)
+{
+	return le16_get_bits(desc->u.wcn7850.msdu_end.info5,
+			     RX_MSDU_END_INFO5_DA_IS_MCBC);
+}
+
 const struct hal_ops hal_wcn7850_ops = {
 	.rx_desc_get_first_msdu = ath12k_hw_wcn7850_rx_desc_get_first_msdu,
 	.rx_desc_get_last_msdu = ath12k_hw_wcn7850_rx_desc_get_last_msdu,
@@ -1151,7 +1164,7 @@ const struct hal_ops hal_wcn7850_ops = {
 	.rx_desc_get_msdu_end_offset = ath12k_hw_wcn7850_rx_desc_get_msdu_end_offset,
 	.rx_desc_mac_addr2_valid = ath12k_hw_wcn7850_rx_desc_mac_addr2_valid,
 	.rx_desc_mpdu_start_addr2 = ath12k_hw_wcn7850_rx_desc_mpdu_start_addr2,
-	.rx_desc_is_mcbc = ath12k_hw_wcn7850_rx_desc_is_mcbc,
+	.rx_desc_ra_is_mcbc = ath12k_hw_wcn7850_rx_desc_ra_is_mcbc,
 	.rx_desc_get_dot11_hdr = ath12k_hw_wcn7850_rx_desc_get_dot11_hdr,
 	.rx_desc_get_crypto_header = ath12k_hw_wcn7850_rx_desc_get_crypto_hdr,
 	.create_srng_config = ath12k_hal_srng_create_config_wcn7850,
@@ -1161,6 +1174,7 @@ const struct hal_ops hal_wcn7850_ops = {
 	.dp_rx_h_ip_cksum_fail = ath12k_hw_wcn7850_dp_rx_h_ip_cksum_fail,
 	.dp_rx_h_is_decrypted = ath12k_hw_wcn7850_dp_rx_h_is_decrypted,
 	.dp_rx_h_mpdu_err = ath12k_hw_wcn7850_dp_rx_h_mpdu_err,
+	.rx_desc_da_is_mcbc = ath12k_hw_wcn7850_rx_desc_da_is_mcbc,
 };
 
 static int ath12k_hal_alloc_cont_rdp(struct ath12k_base *ab)
--- a/drivers/net/wireless/ath/ath12k/hal.h
+++ b/drivers/net/wireless/ath/ath12k/hal.h
@@ -1128,7 +1128,7 @@ struct hal_ops {
 	u32 (*rx_desc_get_msdu_end_offset)(void);
 	bool (*rx_desc_mac_addr2_valid)(struct hal_rx_desc *desc);
 	u8* (*rx_desc_mpdu_start_addr2)(struct hal_rx_desc *desc);
-	bool (*rx_desc_is_mcbc)(struct hal_rx_desc *desc);
+	bool (*rx_desc_ra_is_mcbc)(struct hal_rx_desc *desc);
 	void (*rx_desc_get_dot11_hdr)(struct hal_rx_desc *desc,
 				      struct ieee80211_hdr *hdr);
 	void (*rx_desc_get_crypto_header)(struct hal_rx_desc *desc,
@@ -1145,6 +1145,7 @@ struct hal_ops {
 	u16 (*rxdma_ring_wmask_rx_mpdu_start)(void);
 	u32 (*rxdma_ring_wmask_rx_msdu_end)(void);
 	u8 (*rx_desc_get_msdu_src_link_id)(struct hal_rx_desc *desc);
+	bool (*rx_desc_da_is_mcbc)(struct hal_rx_desc *desc);
 };
 
 extern const struct hal_ops hal_qcn9274_ops;
